lib/livebook_web/live/hub_live.ex

```diff
@@ -0,0 +1,321 @@
+          autocomplete: "off"
+        ) %>
+      </div>
+
+      <%= if length(@hubs) > 0 do %>
+        <div class="flex flex-col space-y-1">
+          <h3 class="text-gray-800 font-semibold">
+            Organization
+          </h3>
+          <%= select(f, :organization, @hubs, class: "input", required: true) %>
+        </div>
+
+        <div class="grid grid-cols-1 md:grid-cols-2 gap-3">
+          <div class="flex flex-col space-y-1">
+            <h3 class="text-gray-800 font-semibold">
+              Name
+            </h3>
+            <%= text_input(f, :name, value: @data["name"], class: "input", required: true) %>
+          </div>
+
+          <div class="flex flex-col space-y-1">
+            <h3 class="text-gray-800 font-semibold">
+              Color
+            </h3>
+
+            <div class="flex space-x-4 items-center">
+              <div
+                class="border-[3px] rounded-lg p-1 flex justify-center items-center"
+                style={"border-color: #{@data["hex_color"]}"}
+              >
+                <div class="rounded h-5 w-5" style={"background-color: #{@data["hex_color"]}"}></div>
+              </div>
+              <div class="relative grow">
+                <%= text_input(f, :hex_color,
+                  value: @data["hex_color"],
+                  class: "input",
+                  spellcheck: "false",
+                  required: true,
+                  maxlength: 7
+                ) %>
+                <button
+                  class="icon-button absolute right-2 top-1"
+                  type="button"
+                  phx-click="randomize_color"
+                >
+                  <.remix_icon icon="refresh-line" class="text-xl" />
+                </button>
+              </div>
+            </div>
+          </div>
+        </div>
+
+        <%= submit("Save", class: "button-base button-blue", phx_disable_with: "Saving...") %>
+      <% end %>
+    </.form>
+    """
+  end
+
+  @impl true
+  def handle_params(%{"id" => id}, _url, socket) do
+    hub = Hubs.hub_by_id!(id)
+    hubs = [hub]
+
+    data =
+      case hub.type do
+        "fly" ->
+          %{
+            "organization" => hub.id,
+            "token" => hub.token,
+            "name" => hub.name,
+            "hex_color" => hub.color
+          }
+      end
+
+    opts = select_hub_options(hubs, data["organization"])
+
+    {:noreply,
+     assign(socket,
+       operation: :edit,
+       selected_hub_service: hub.type,
+       data: data,
+       hubs: hubs,
+       hub_options: opts
+     )}
+  end
+
+  def handle_params(_params, _url, socket) do
+    {:noreply, assign(socket, operation: :new)}
+  end
+
+  @impl true
+  def handle_event("select_hub_service", %{"value" => service}, socket) do
+    {:noreply, assign(socket, selected_hub_service: service)}
+  end
+
+  def handle_event("fetch_hubs", %{"fly" => %{"token" => token}}, socket) do
+    case HubProvider.fetch_hubs(%Fly{token: token}) do
```

Comment:
The creation of the Fly Hub itself does not go through a provider. Remember that each "provider" (Fly, Enterprise) is going to have a distinct setup with different properties, so we shouldn't try to unify them under a single abstraction.

My suggestion is to get rid of HubProvider here and call `Fly.fetch_organizations` or `Fly.fetch_applications` accordingly. We won't try to normalize the access in this PR yet. :)

Reply:
In fact, my suggestion is to use distinct names to avoid confusion. You are not fetching hubs here, the hubs only exist after they are saved (i.e. all hubs are saved_hubs).

Also, perhaps to simplify the logic in the future, I would move the Fly specific logic to a live component. I.e. Once you click on "Fly" to create a Hub, it renderes a live component with the next steps and we perform everything from the component!